### DISCUSSION TITLE
My LEDS don't show the expected effects

### DISCUSSION THREAD

**Discussion by helenfromyorkshire** [PlummersSoftwareLLC/NightDriverStrip#234] [General]
Hello,

I am interested in using Night Driver to light some LEDs in my garage.

So, I have done the following:

- set up my hardware (M5Stick-plus, an LED strip (85 pixels) and a 5V power supply) as described in the Daves Garage Video 'Laser Cannon'
- created a secrets.h file with my wi-fi details
- 
- edited globals.h	#define LED_PIN0             32  // was 26 ... to match the grove input (G32) on the M5Stick
- edited globals.h	#define MATRIX_WIDTH    85 // my LED strip is 85 pixels
- edited globals.h	#define MATRIX_HEIGHT   1
- 
- edited platformio.ini	upload_port     = COM6 //  in [base]
- edited platformio.ini	monitor_port    = COM6 //  in [base]
- edited platformio.ini	default_envs = spectrum // in [platformio]
- 
- uploaded the code using the 'spectrum' environment

The display on the M5stick shows the Spectrum Analyser which reacts to sound fine and has an IP address.

The LED strip lights up with various patterns (clicking through the effect numbers 1 to 12 in the M5Stick), but doesn't do the effects I would expect.

Choosing each effect has this result (together with the colors in the enclosed image file) :

1/12 Spectrum Standard:
a fixed rainbow pattern
one pixel (#85) flashes red when sound detected
the effect is stable for approximately 7 minutes, then changes to 2/12

2/12 Spectrum Standard:
a fixed rainbow pattern
pixels 69 to 85 flash when sound detected

3/12 Spectrum Standard:
a fixed rainbow pattern
pixels 69 to 85 flash when sound detected

4/12 Spectrum Standard:
a fixed rainbow pattern
pixels 81 to 85 flash when sound detected
the effect is stable for approximately 4 minutes, then changes to 1/12

5/12 Spectrum Standard:
pixels 1 to 49 are blue, with reduced intensity
pixels 50 to 85 flash when sound detected

6/12 GhostWave:
pixels 42 to 45 are red
all other pixels are off
the effect resets to 1/12 after 5 seconds

7/12 Spectrum USA:
pixels 1 to 80 show the colors of the US flag
pixels 81 to 85 flash when sound detected

8/12 GhostWave Rainbow:
pixels 43 to 44 are red
all other pixels are off
the effect resets to 1/12 after 5 seconds

9/12 Spectrum Fade:
all pixels react to sound  ... a moving spectrum from the ends towards the center

10/12 GhostWave Blue:
pixels 43 to 44 are red
all other pixels are off with no reaction to sound
the effect resets to 1/12 after 5 seconds

11/12 Spectrum Standard:
a fixed rainbow pattern
pixels 69 to 85 flash when sound detected

12/12 GhostWave Blue:
pixels 43 to 44 are red
all other pixels are off with no reaction to sound
the effect resets to 1/12 after 5 seconds

I hope someone can help me get my lights going.

Thank you for your help. Helen.

![M5stick-plus](https://user-images.githubusercontent.com/126163493/223702196-4b719765-a7c7-4f47-9077-015102e25dc7.png)

**Comment by rbergen** [comment]
You say the effects you're seeing are not the ones you would expect. I'm curious to know what are the effects you would expect?

I think the root issue here is that the Spectrum effects are matrix effects. That means they're intended and programmed to be used on a two-dimensional oblong of LEDs, not on single-line strips. In fact, the Spectrum effects are specifically designed for a matrix consisting of 48x16 LEDs. That matrix is usually composed of three 16x16 panels that are chained in sequence. For reference, I've added a picture of my setup running the Spectrum project.

![48x16 Spectrum setup](https://user-images.githubusercontent.com/10425218/223710310-64220656-6590-4b2a-9a06-a4ffcc5175da.jpg)

**Reply by helenfromyorkshire** [reply]
Hi Rutger ... thank you for replying so quickly.

You have exposed my inexperience in doing this sort of thing.

So, could you help me with a simpler task?  I have watched Dave's Laser Cannon video again and I want to simply get that to work.

When I change the environment to 'laserline' and upload that, I get an unstable display on the m5stick (see enclosed video), half my LED strip is blue and the output from the serial monitor in VSCode shows something is corrupt (see enclosed text file).

I would be very grateful if you could have a look and give me some pointers on how to fix things.

Thanks again Rutger,
Helen
[VSCode Output.txt](https://github.com/PlummersSoftwareLLC/NightDriverStrip/files/10921094/VSCode.Output.txt)

https://user-images.githubusercontent.com/126163493/223729160-4aa88dfb-d954-4ca2-8a92-d4abc39bc2c3.mov

**Comment by davepl** [comment]
Sounds like you’re very close, but Spectrum wouldn’t be appropriate for a strip unless that strip were woven into a boustrophedon matrix.  Which is really just an excuse to flex the word boustrophedon, but I digress.

I think you have everything working, now you just need to write or select effects that do what you want and place them in the SPECTRUM section of the effects table in effects.cpp (or create a new config, like HELEN or YORKSHIRE, etc).

So if you put a single entry like

        new RainbowFillEffect(48, 0),

You’d get a single rainbow effect across your 85 LEDs.

Hope that’s what you were asking, if not, type more slowly and I’ll try again :-)

Cheers,
Dave

**Reply by helenfromyorkshire** [reply]
Thanks a lot Dave ... I am trying that now, Helen.

**Comment by helenfromyorkshire** [comment]
Thank you very much Dave.  I seem to have more control of my LEDs now.

I bought the LEDs and the m5stick to try and have a 'laser cannon' in my garage, so could you please help me get another step further?

I've commented out all the effects in SPECTRUM (in effects.cpp) except "Spectrum Fade" and added 'new LaserLineEffect (500,20)' at the bottom of the section.

They both work, but (there's always a but) ...

LaserLineEffect is selectable using the button on the side of the m5stick, but only works for a few seconds, before returning to Spectrum Fade.  I think it resets the m5stick (see enclosed serial monitor text).

> How can I make LaserLineEffect more persistent?
> Can I alter something that makes both of these effects fire/react to lower volumes?

Thanks again for your help.
Helen

[VSCode Output 2.txt](https://github.com/PlummersSoftwareLLC/NightDriverStrip/files/10922166/VSCode.Output.2.txt)

**Comment by davepl** [comment]
If your serial output includes any info on why the effect stopped or crashed, please pass it along.

There are two values you can tweak for audio sensitivity:  NOISE_LEVEL is how loud any particular sample needs to be to register, and NOISE_FLOOR is the total amount of noise that is ignored.  Try reducing those!

Thanks
Dave

**Reply by rbergen** [reply]
@davepl I think the following lines in the file @helenfromyorkshire attached earlier show the moment that the problem occurs:

```text
E (87064) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (87064) task_wdt:  - Idle0 (CPU 0)
E (87064) task_wdt: Tasks currently running:
E (87064) task_wdt: CPU 0: Screen Loop
E (87064) task_wdt: CPU 1: Idle1
E (87064) task_wdt: Aborting.

abort() was called at PC 0x400ff390 on core 0

Backtrace: 0x40084bf9:0x3ffbec9c |<-CORRUPTED

  #0  0x40084bf9:0x3ffbec9c in panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:402   
```

I'm just spitballing here, but could it be because LaserLine is not a spectrum effect, and hence doesn't combine with the spectrum analyzer shown on the M5Stick's built-in display?

**Comment by davepl** [comment]
It’s a bit of a long shot, but depending on which code you have, check the fn DrawLoopTaskEntry.

Near the end there may be a yield() instruction.  If so, change that to delay(5) and see if that fixes it.

I think under FreeRTOS, even if you yield, if your thread priority is higher than idle and you demand all the cpu, you get it, and the watchdog bites you as punishment.

- Dave
